test/functional/test_framework/test_framework.py

```diff
@@ -251,7 +260,29 @@ def main(self):
         logging.shutdown()
         if cleanup_tree_on_exit:
             shutil.rmtree(self.options.tmpdir)
-        sys.exit(exit_code)
+
+        for h in list(self.log.handlers):
+            self.log.removeHandler(h)
+        self.nodes.clear()
+        return exit_code
+
+    def handle_exception(self, e):
+        # Restore traceback from exception object.
+        tb = traceback.format_exception(type(e), e, e.__traceback__)
```

Comment:
Not Ideal, I agree. It seems the logger will only capture the traceback if error logging is called directly inside `except` (see https://stackoverflow.com/a/5191885 and https://docs.python.org/3/library/logging.html#logging.error). Since we are invoking error/exception logging elsewhere, we reconstruct the traceback string from the exception object.